internal/modelcontroller/cache.go

```diff
@@ -0,0 +1,469 @@
+package modelcontroller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	kubeaiv1 "github.com/substratusai/kubeai/api/v1"
+	"github.com/substratusai/kubeai/internal/k8sutils"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type PVCModelAnnotationValue struct {
+	UID       string    `json:"uid"`
+	Timestamp time.Time `json:"timestamp"`
+}
+
+func (r *ModelReconciler) reconcileCache(ctx context.Context, model *kubeaiv1.Model, cfg ModelConfig) (ctrl.Result, error) {
+	if model.Status.Cache == nil {
+		model.Status.Cache = &kubeaiv1.ModelStatusCache{}
+	}
+
+	modelDeleted := model.DeletionTimestamp != nil
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	var pvcExists bool
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: model.Namespace,
+		Name:      cachePVCName(model, cfg),
+	}, pvc); err != nil {
+		if apierrors.IsNotFound(err) {
+			pvcExists = false
+		} else {
+			return ctrl.Result{}, fmt.Errorf("getting cache PVC: %w", err)
+		}
+	} else {
+		pvcExists = true
+	}
+
+	// Create PVC if not exists.
+	if !pvcExists {
+		if !modelDeleted {
+			pvc = r.cachePVCForModel(model, cfg)
+			// TODO: Set controller reference on PVC for 1:1 Model to PVC situations
+			// such as Google Hyperdisk ML.
+			//if err := controllerutil.SetControllerReference(model, pvc, r.Scheme); err != nil {
+			//	return ctrl.Result{}, fmt.Errorf("setting controller reference on pvc: %w", err)
+			//}
+			if err := r.Create(ctx, pvc); err != nil {
+				return ctrl.Result{}, fmt.Errorf("creating cache PVC: %w", err)
+			}
+		}
+	}
+
+	// Caches that are shared across multiple Models require model-specific cleanup.
+	if cfg.CacheProfile.SharedFilesystem != nil {
+		if controllerutil.AddFinalizer(model, kubeaiv1.ModelCacheEvictionFinalizer) {
```

Comment:
Would it make sense to check for the finalizer first, to not add duplicates?

Reply:
The AddFinalizer() func returns a bool indicating whether it mutated the finalizers